### PR TITLE
Add Wilder Accumulative Swing Index

### LINKED ISSUE #5387: Adds Wilder Accumulation Swing Index Indicator
#### Expected Behavior
[Wilder Accumulation Swing Index](https://www.investopedia.com/terms/a/asi.asp) indicator is natively supported by Lean

#### Actual Behavior
Lean does not support ASI.

#### Potential Solution
Implement it.

#### Checklist
- [x] I have completely filled out this template
- [x] I have confirmed that this issue exists on the current `master` branch
- [x] I have confirmed that this is not a duplicate issue by searching [issues](https://github.com/QuantConnect/Lean/issues)

### PR DESCRIPTION
**Pull request by david-acker**
#### Description
Added the <code>WilderAccumulativeSwingIndex</code> (ASI) indicator, including the underlying <code>WilderSwingIndex</code> (SI) indicator used for its calculation, with the information provided in Welles Wilder's book <i>New Concepts in Technical Trading Systems</i> (1978). 

#### Related Issue
Closes #5387

#### Motivation and Context
Requested indicator.

#### Requires Documentation Change
No.

#### How Has This Been Tested?
Successfully passed all tests in WilderAccumulativeSwingIndexTests.cs and WilderSwingIndexTests.cs, using test data from "spy_asi.csv" and "spy_si.csv" respectively. Test data was sourced from the "spy_ad.txt" file included in the TestData folder. Additionally, no errors were thrown when <code>WilderAccumulativeSwingIndex</code> and <code>WilderSwingIndex</code> were added to IndicatorWarmupAlgorithm.cs.

Update: 4-13-21
Test data in "spy_asi.csv" and "spy_si.csv" was replaced with data sourced from TradingView.

#### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Refactor (non-breaking change which improves implementation)
- [ ] Performance (non-breaking change which improves performance. Please add associated performance test and results)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Non-functional change (xml comments/documentation/etc)

#### Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the **CONTRIBUTING** [document](https://github.com/QuantConnect/Lean/blob/master/CONTRIBUTING.md).
- [x] I have added tests to cover my changes. 
- [x] All new and existing tests passed.
- [x] My branch follows the naming convention `bug-<issue#>-<description>` or `feature-<issue#>-<description>`

### CODE REVIEW

File: Algorithm/QCAlgorithm.Indicators.cs [hunk 1]
```diff
@@ -1880,6 +1880,52 @@ public WilderMovingAverage WWMA(Symbol symbol, int period, Resolution? resolutio
...
+        /// The indicator will be automatically updated on the given resolution.
+        /// </summary>
+        /// <param name="symbol">The symbol whose SI we want</param>
+        /// <param name="limitMove">The maximum daily change in price for the SI</param>
+        /// <param name="resolution">The resolution</param>
+        /// <returns>The WilderSwingIndex for the given parameters</returns>
+        /// <remarks>SI for Wilder Swing Index</remarks>
+        public WilderSwingIndex SI(Symbol symbol, decimal limitMove, Resolution? resolution = Resolution.Daily)
+        {
+            var name = CreateIndicatorName(symbol, $"ASI", resolution);
```
Comment (jaredbroad) [inline Algorithm/QCAlgorithm.Indicators.cs:1894]: Can you please rename the indicator name "SI"

File: Indicators/WilderAccumulativeSwingIndex.cs [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+    ///     <item>
+    ///       <term>SIₜ</term>
+    ///       <description>
+    ///         The <see cref="WilderSwingIndex"/> calculated for the current period.
+    ///       </description>
+    ///     </item>
+    ///   </list>
+    /// </para>
+    /// </summary>
+    /// <seealso cref="WilderSwingIndex"/>
```
Comment (jaredbroad) [inline Indicators/WilderAccumulativeSwingIndex.cs:44]: 👏 Great comments.

File: Indicators/WilderSwingIndex.cs [hunk 3]
```diff
@@ -0,0 +1,286 @@
...
+    ///         Found by selecting the expression with the largest value and
+    ///         then using the corresponding formula.
+    ///         <para>
+    ///           <i>Expression</i> ⭢ <i>Formula</i>
+    ///           <list type="number">
+    ///             <item>
+    ///               <description>
+    ///                 |Hₜ - Cₜ₋₁|
+    ///                 ⭢
+    ///                 |Hₜ - Cₜ - 0.5 * (Lₜ - Cₜ₋₁) + 0.25 * (Cₜ₋₁ - Oₜ₋₁)|
```
Comment (jaredbroad) [inline Indicators/WilderSwingIndex.cs:52]: Does this pass XML spec; I'm not sure it'll accept special characters

File: Tests/QuantConnect.Tests.csproj [hunk 4]
```diff
@@ -304,6 +304,12 @@
     <None Update="TestData\spy_arima.csv">
       <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
     </None>
+    <None Update="TestData\spy_si.csv">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
+    <None Update="TestData\spy_asi.csv">
+      <CopyToOutputDirectory>PreserveNewest</CopyToOutputDirectory>
+    </None>
```
Comment (jaredbroad) [inline Tests/QuantConnect.Tests.csproj:312]: > Test data was sourced from the "spy_ad.txt" file included in the TestData folder.

Where was the original source of the data ( i.e. a third party source or self-generated? )  We need all indicators to be vetted against a third-party source as the community routinely sanity checks the indicator results, its always good to know in advance its been compared and passed.

**Review by jaredbroad — CHANGES_REQUESTED** [review]
Thank you for the PR and welcome to lean @david-acker! Minor comments but close to merging.

File: Algorithm/QCAlgorithm.Indicators.cs (hunk 1 above)
Comment (david-acker) [reply Algorithm/QCAlgorithm.Indicators.cs]: Fixed in 87c8991

File: Indicators/WilderSwingIndex.cs (hunk 3 above)
Comment (david-acker) [reply Indicators/WilderSwingIndex.cs]: Fixed in 0f22f76.

File: Tests/QuantConnect.Tests.csproj (hunk 4 above)
Comment (david-acker) [reply Tests/QuantConnect.Tests.csproj:312]: Fixed in fe76c48 and f016058. After a brief look around, I couldn't find any info on the original source of "spy_ad.txt", so I simply replaced the test data sets for SI and ASI with data sourced from TradingView.

In the process, I also discovered that the R value for SI was being calculated incorrectly. This issue was fixed in 9b41f50.

File: Indicators/WilderAccumulativeSwingIndex.cs (hunk 2 above)
Comment (david-acker) [reply Indicators/WilderAccumulativeSwingIndex.cs:44]: Thank you!

**Review by jaredbroad — APPROVED** [review]

**Review by Martin-Molinero — APPROVED** [review]

**Comment by Martin-Molinero** [comment]
Awesome PR @david-acker!

**Pull request merged**